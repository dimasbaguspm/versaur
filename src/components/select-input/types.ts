import type { SelectHTMLAttributes, ReactNode } from 'react'

/**
 * Props for the SelectInput component
 */
export interface SelectInputProps
  extends Omit<SelectHTMLAttributes<HTMLSelectElement>, 'size'> {
  /**
   * Visual style variant supporting Versaur color system
   * Core variants: primary (coral), secondary (sage), tertiary (mist), ghost (slate), neutral (light gray)
   * Semantic variants: success, info, warning, danger
   * Each variant supports outline form for flexible design expression
   */
  variant?:
    | 'primary'
    | 'primary-outline'
    | 'secondary'
    | 'secondary-outline'
    | 'tertiary'
    | 'tertiary-outline'
    | 'ghost'
    | 'ghost-outline'
    | 'neutral'
    | 'neutral-outline'
    | 'success'
    | 'success-outline'
    | 'info'
    | 'info-outline'
    | 'warning'
    | 'warning-outline'
    | 'danger'
    | 'danger-outline'
  /**
   * Label text to display above the select
   */
  label: ReactNode
  /**
   * Helper text to display below the select
   */
  helperText?: ReactNode
  /**
   * Error message for invalid state
   */
  error?: ReactNode
  /**
   * Placeholder text for when no option is selected
   */
  placeholder?: string
}
