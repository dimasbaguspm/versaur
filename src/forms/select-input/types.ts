import type {
  SelectHTMLAttributes,
  OptionHTMLAttributes,
  ReactNode,
} from 'react'

/**
 * Props for the SelectInput component
 */
export interface SelectInputProps
  extends Omit<SelectHTMLAttributes<HTMLSelectElement>, 'size'> {
  /**
   * Label text to display above the select
   */
  label?: ReactNode
  /**
   * Helper text to display below the select
   */
  helperText?: ReactNode
  /**
   * Error message for invalid state
   */
  error?: ReactNode
  /**
   * Placeholder text for when no option is selected
   */
  placeholder?: string
  /**
   * Whether the select is read-only
   */
  readOnly?: boolean
}

/**
 * Props for the SelectOption component
 */
export interface SelectOptionProps
  extends OptionHTMLAttributes<HTMLOptionElement> {
  /**
   * Option value (required)
   */
  value: string
  /**
   * Option label/content (required)
   */
  children: ReactNode
}

/**
 * Props for the SelectOptionGroup component
 */
export interface SelectOptionGroupProps
  extends React.OptgroupHTMLAttributes<HTMLOptGroupElement> {
  /**
   * Group label (required)
   */
  label: string
  /**
   * Options within the group (required)
   */
  children: ReactNode
}
