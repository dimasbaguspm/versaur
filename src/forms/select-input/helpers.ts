import { cva } from '@/utils/variants'

export const selectInputVariants = cva(
  'block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:pointer-events-none disabled:bg-gray-50 appearance-none bg-no-repeat bg-right pr-10',
  {
    variants: {
      state: {
        default:
          'border-primary/30 text-foreground focus:border-primary focus:ring-primary/20',
        error:
          'border-danger bg-danger/5 text-foreground focus:border-danger focus:ring-danger/20',
        readOnly:
          'border-gray-300 bg-gray-50 text-foreground cursor-default focus:ring-0 focus:border-gray-300',
      },
    },
    defaultVariants: {
      state: 'default',
    },
  }
)
