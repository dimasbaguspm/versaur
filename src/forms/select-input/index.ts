export { SelectInput } from './select-input'
export type {
  SelectInputProps,
  SelectOptionProps,
  SelectOptionGroupProps,
} from './types'
