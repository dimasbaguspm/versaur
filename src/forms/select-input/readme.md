# SelectInput

A styled select dropdown component that follows browser standards and provides a professional appearance with Versaur's harmonious color palette.

## Features

- **Semantic Variants**: Support for all Versaur color variants (primary, secondary, tertiary, ghost, neutral) and semantic states (success, info, warning, danger)
- **Outline Variants**: Each variant includes an outline version for subtle styling
- **Accessibility**: Full ARIA support with proper labeling and error states
- **Error Handling**: Built-in error state with validation messages
- **Helper Text**: Optional helper text for additional guidance
- **Placeholder Support**: Optional placeholder text for better UX
- **Disabled State**: Proper disabled styling and behavior
- **Custom Dropdown Arrow**: Consistent arrow styling across browsers
- **Form Integration**: Works seamlessly with forms and validation libraries

## Usage

```tsx
import { SelectInput } from '@versaur/ui'

// Basic usage
<SelectInput
  label="Choose your country"
  placeholder="Select a country..."
>
  <option value="us">United States</option>
  <option value="ca">Canada</option>
  <option value="uk">United Kingdom</option>
</SelectInput>

// With helper text
<SelectInput
  label="Select your plan"
  helperText="You can change this later in settings"
  placeholder="Choose a plan..."
>
  <option value="free">Free Plan</option>
  <option value="pro">Pro Plan</option>
</SelectInput>

// Error state
<SelectInput
  label="Required field"
  error="Please select an option"
  placeholder="Make a selection..."
>
  <option value="option1">Option 1</option>
  <option value="option2">Option 2</option>
</SelectInput>

// With option groups
<SelectInput
  label="Select a vehicle"
  placeholder="Choose your vehicle..."
>
  <optgroup label="Cars">
    <option value="sedan">Sedan</option>
    <option value="suv">SUV</option>
  </optgroup>
  <optgroup label="Motorcycles">
    <option value="sport">Sport Bike</option>
    <option value="cruiser">Cruiser</option>
  </optgroup>
</SelectInput>
```

## Props

The component extends all standard HTML `select` attributes and adds the following:

- `variant`: Visual style variant (primary, secondary, tertiary, ghost, neutral, success, info, warning, danger, plus outline versions)
- `label`: Required label text displayed above the select
- `helperText`: Optional helper text displayed below the select
- `error`: Error message that overrides helper text and sets error styling
- `placeholder`: Optional placeholder text for the default option

## Accessibility

- Proper ARIA attributes (`aria-invalid`, `aria-disabled`)
- Semantic HTML with label association
- Error messages use `role="alert"` for screen reader announcements
- Focus management and keyboard navigation follow browser standards
