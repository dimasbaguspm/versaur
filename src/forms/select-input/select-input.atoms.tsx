import React from 'react'
import type { SelectOptionProps, SelectOptionGroupProps } from './types'

/**
 * SelectOption component - an atom for SelectInput
 *
 * Wraps native HTML option element with proper typing
 * Part of the SelectInput compound pattern
 */
export const SelectOption = React.forwardRef<
  HTMLOptionElement,
  SelectOptionProps
>(({ value, children, ...props }, ref) => {
  return (
    <option ref={ref} value={value} {...props}>
      {children}
    </option>
  )
})

SelectOption.displayName = 'SelectOption'

/**
 * SelectOptionGroup component - an atom for SelectInput
 *
 * Wraps native HTML optgroup element with proper typing
 * Part of the SelectInput compound pattern
 */
export const SelectOptionGroup = React.forwardRef<
  HTMLOptGroupElement,
  SelectOptionGroupProps
>(({ label, children, ...props }, ref) => {
  return (
    <optgroup ref={ref} label={label} {...props}>
      {children}
    </optgroup>
  )
})

SelectOptionGroup.displayName = 'SelectOptionGroup'
