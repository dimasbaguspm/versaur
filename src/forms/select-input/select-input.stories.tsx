/**
 * SelectInput stories for Storybook
 *
 * Demonstrates core functionality: basic select, helper text, error states, and disabled/read-only states
 * A simple select dropdown component that aligns with native HTML behavior
 */
import type { Meta, StoryObj } from '@storybook/react'
import { SelectInput } from './select-input'

const meta: Meta<typeof SelectInput> = {
  title: 'Forms/SelectInput',
  component: SelectInput,
  tags: ['autodocs'],
}
export default meta

type Story = StoryObj<typeof SelectInput>

/**
 * Default select input with label and placeholder using SelectInput.Option compound pattern
 */
export const Default: Story = {
  args: {
    label: 'Country',
    placeholder: 'Select a country',
    children: (
      <>
        <SelectInput.Option value='us'>United States</SelectInput.Option>
        <SelectInput.Option value='ca'>Canada</SelectInput.Option>
        <SelectInput.Option value='uk'>United Kingdom</SelectInput.Option>
        <SelectInput.Option value='de'>Germany</SelectInput.Option>
        <SelectInput.Option value='fr'>France</SelectInput.Option>
      </>
    ),
  },
}

/**
 * Select input with helper text below for additional guidance
 */
export const WithHelperText: Story = {
  args: {
    label: 'Department',
    placeholder: 'Choose department...',
    helperText: 'Select your primary department',
    children: (
      <>
        <SelectInput.Option value='engineering'>Engineering</SelectInput.Option>
        <SelectInput.Option value='design'>Design</SelectInput.Option>
        <SelectInput.Option value='marketing'>Marketing</SelectInput.Option>
        <SelectInput.Option value='sales'>Sales</SelectInput.Option>
        <SelectInput.Option value='support'>Support</SelectInput.Option>
      </>
    ),
  },
}

/**
 * Select input in error state with validation message
 */
export const WithError: Story = {
  args: {
    label: 'Plan',
    placeholder: 'Choose a plan...',
    error: 'Please select a plan to continue',
    children: (
      <>
        <SelectInput.Option value='free'>Free Plan</SelectInput.Option>
        <SelectInput.Option value='basic'>
          Basic Plan - $9/month
        </SelectInput.Option>
        <SelectInput.Option value='pro'>
          Pro Plan - $19/month
        </SelectInput.Option>
        <SelectInput.Option value='enterprise'>
          Enterprise Plan - $99/month
        </SelectInput.Option>
      </>
    ),
  },
}

/**
 * Required field with asterisk indicator
 */
export const Required: Story = {
  args: {
    label: 'Experience Level',
    placeholder: 'Select experience...',
    required: true,
    helperText: 'This field is required',
    children: (
      <>
        <SelectInput.Option value='junior'>
          Junior (0-2 years)
        </SelectInput.Option>
        <SelectInput.Option value='mid'>
          Mid-level (2-5 years)
        </SelectInput.Option>
        <SelectInput.Option value='senior'>
          Senior (5-10 years)
        </SelectInput.Option>
        <SelectInput.Option value='lead'>Lead (10+ years)</SelectInput.Option>
      </>
    ),
  },
}

/**
 * Disabled select input that cannot be interacted with
 */
export const Disabled: Story = {
  args: {
    label: 'Unavailable Options',
    placeholder: 'Options not available',
    disabled: true,
    helperText: 'This selection is currently disabled',
    children: (
      <>
        <SelectInput.Option value='option1'>Option 1</SelectInput.Option>
        <SelectInput.Option value='option2'>Option 2</SelectInput.Option>
        <SelectInput.Option value='option3'>Option 3</SelectInput.Option>
      </>
    ),
  },
}

/**
 * Read-only select input that displays data but cannot be modified
 */
export const ReadOnly: Story = {
  args: {
    label: 'Account Type',
    readOnly: true,
    defaultValue: 'premium',
    helperText: 'This field is read-only and displays reference information',
    children: (
      <>
        <SelectInput.Option value='free'>Free</SelectInput.Option>
        <SelectInput.Option value='basic'>Basic</SelectInput.Option>
        <SelectInput.Option value='premium'>Premium</SelectInput.Option>
        <SelectInput.Option value='enterprise'>Enterprise</SelectInput.Option>
      </>
    ),
  },
}

/**
 * Select input with grouped options using SelectInput.OptionGroup for better organization
 */
export const WithOptionGroups: Story = {
  args: {
    label: 'Vehicle',
    placeholder: 'Choose your vehicle...',
    helperText: 'Options are grouped by category',
    children: (
      <>
        <SelectInput.OptionGroup label='Cars'>
          <SelectInput.Option value='sedan'>Sedan</SelectInput.Option>
          <SelectInput.Option value='suv'>SUV</SelectInput.Option>
          <SelectInput.Option value='hatchback'>Hatchback</SelectInput.Option>
          <SelectInput.Option value='coupe'>Coupe</SelectInput.Option>
        </SelectInput.OptionGroup>
        <SelectInput.OptionGroup label='Motorcycles'>
          <SelectInput.Option value='sport'>Sport Bike</SelectInput.Option>
          <SelectInput.Option value='cruiser'>Cruiser</SelectInput.Option>
          <SelectInput.Option value='touring'>Touring</SelectInput.Option>
          <SelectInput.Option value='dirt'>Dirt Bike</SelectInput.Option>
        </SelectInput.OptionGroup>
        <SelectInput.OptionGroup label='Trucks'>
          <SelectInput.Option value='pickup'>Pickup Truck</SelectInput.Option>
          <SelectInput.Option value='van'>Van</SelectInput.Option>
          <SelectInput.Option value='box-truck'>Box Truck</SelectInput.Option>
        </SelectInput.OptionGroup>
      </>
    ),
  },
}
