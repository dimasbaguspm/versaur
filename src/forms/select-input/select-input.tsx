import React from 'react'
import { ChevronDown } from 'lucide-react'
import { cn } from '@/utils/cn'
import { Icon } from '@/primitive/icon'
import { selectInputVariants } from './helpers'
import { SelectOption, SelectOptionGroup } from './select-input.atoms'
import type { SelectInputProps } from './types'

/**
 * SelectInput component for Versaur UI
 *
 * A simple, accessible select dropdown that aligns with native HTML select element
 * Follows browser standards and accessibility best practices with proper ARIA attributes
 */
const SelectInputRoot = React.forwardRef<HTMLSelectElement, SelectInputProps>(
  (
    {
      label,
      helperText,
      error,
      className,
      disabled,
      readOnly,
      id,
      required,
      placeholder,
      children,
      onChange,
      ...props
    },
    ref
  ) => {
    const generatedId = React.useId()
    const selectId = id || generatedId
    const helperTextId = `${selectId}-helper`
    const errorId = `${selectId}-error`
    const hasError = Boolean(error)

    // Build aria-describedby based on what's present
    const ariaDescribedBy = React.useMemo(() => {
      const ids: string[] = []
      if (hasError) {
        ids.push(errorId)
      } else if (helperText) {
        ids.push(helperTextId)
      }
      return ids.length > 0 ? ids.join(' ') : undefined
    }, [hasError, helperText, errorId, helperTextId])

    return (
      <div className={cn('w-full', className)}>
        {label && (
          <label
            htmlFor={selectId}
            className='block text-sm font-medium text-foreground mb-2'
          >
            {label}
            {required && (
              <span className='text-danger ml-1' aria-label='required'>
                *
              </span>
            )}
          </label>
        )}
        <div className='relative w-full'>
          <select
            ref={ref}
            id={selectId}
            required={required}
            disabled={disabled || readOnly}
            aria-invalid={hasError}
            aria-describedby={ariaDescribedBy}
            aria-errormessage={hasError ? errorId : undefined}
            onChange={onChange}
            className={cn(
              selectInputVariants({
                state: hasError ? 'error' : readOnly ? 'readOnly' : 'default',
              }),
              'h-9 pl-3'
            )}
            {...props}
          >
            {placeholder && (
              <option value='' disabled hidden>
                {placeholder}
              </option>
            )}
            {children}
          </select>
          {/* Custom dropdown arrow with flip transition */}
          <div
            className={cn(
              'absolute inset-y-0 right-0 flex items-center pr-3 pointer-events-none transition-transform duration-200'
            )}
          >
            <Icon as={ChevronDown} size='sm' color='gray' aria-hidden='true' />
          </div>
        </div>
        {hasError && (
          <div
            id={errorId}
            className='mt-1 text-sm text-danger'
            role='alert'
            aria-live='polite'
          >
            {error}
          </div>
        )}
        {!hasError && helperText && (
          <div id={helperTextId} className='mt-1 text-sm text-gray-600'>
            {helperText}
          </div>
        )}
      </div>
    )
  }
)

SelectInputRoot.displayName = 'SelectInput'

export const SelectInput = Object.assign(SelectInputRoot, {
  Option: SelectOption,
  OptionGroup: SelectOptionGroup,
})
