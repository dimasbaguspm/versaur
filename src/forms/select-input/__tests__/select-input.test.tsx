import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../select-input.stories'

const { Default, WithError, WithHelperText, Required, Disabled, ReadOnly } =
  composeStories(stories)

describe('SelectInput', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders with label and options', () => {
    render(<Default />)

    expect(screen.getByLabelText('Country')).toBeInTheDocument()
    expect(screen.getByRole('combobox')).toBeInTheDocument()
  })

  it('displays helper text when provided', () => {
    render(<WithHelperText />)

    expect(
      screen.getByText('Select your primary department')
    ).toBeInTheDocument()
  })

  it('displays error message in error state', () => {
    render(<WithError />)

    const select = screen.getByRole('combobox')
    expect(select).toHaveAttribute('aria-invalid', 'true')
    expect(
      screen.getByText('Please select a plan to continue')
    ).toBeInTheDocument()
    expect(screen.getByRole('alert')).toBeInTheDocument()
  })

  it('shows required asterisk when required prop is true', () => {
    render(<Required />)

    const asterisk = screen.getByLabelText('required')
    expect(asterisk).toBeInTheDocument()
    expect(asterisk).toHaveClass('text-danger')

    const select = screen.getByRole('combobox')
    expect(select).toBeRequired()
  })

  it('is disabled when disabled prop is true', () => {
    render(<Disabled />)

    const select = screen.getByRole('combobox')
    expect(select).toBeDisabled()
  })

  it('is disabled when readOnly prop is true', () => {
    render(<ReadOnly />)

    const select = screen.getByRole('combobox')
    expect(select).toBeDisabled()
  })

  it('generates unique ID when not provided', () => {
    render(<Default />)

    const select = screen.getByRole('combobox')
    const label = screen.getByText('Country')

    expect(select).toHaveAttribute('id')
    expect(label).toHaveAttribute('for', select.getAttribute('id'))
  })

  it('uses provided ID when given', () => {
    render(<Default id='custom-select-id' />)

    const select = screen.getByRole('combobox')
    const label = screen.getByText('Country')

    expect(select).toHaveAttribute('id', 'custom-select-id')
    expect(label).toHaveAttribute('for', 'custom-select-id')
  })

  it('shows placeholder option when provided', () => {
    render(<Default />)

    expect(screen.getByText('Select a country')).toBeInTheDocument()
  })

  it('renders all provided options using SelectInput.Option', () => {
    render(<Default />)

    expect(screen.getByText('United States')).toBeInTheDocument()
    expect(screen.getByText('Canada')).toBeInTheDocument()
    expect(screen.getByText('United Kingdom')).toBeInTheDocument()
    expect(screen.getByText('Germany')).toBeInTheDocument()
    expect(screen.getByText('France')).toBeInTheDocument()
  })

  it('does not show helper text when error is present', () => {
    render(<WithError helperText='This should not be visible' />)

    expect(
      screen.queryByText('This should not be visible')
    ).not.toBeInTheDocument()
    expect(
      screen.getByText('Please select a plan to continue')
    ).toBeInTheDocument()
  })

  it('sets aria-describedby when helper text is present', () => {
    render(<WithHelperText />)

    const select = screen.getByRole('combobox')
    const helperId = `${select.id}-helper`
    expect(select).toHaveAttribute('aria-describedby', helperId)
  })

  it('sets aria-describedby and aria-errormessage when error is present', () => {
    render(<WithError />)

    const select = screen.getByRole('combobox')
    const errorId = `${select.id}-error`
    expect(select).toHaveAttribute('aria-describedby', errorId)
    expect(select).toHaveAttribute('aria-errormessage', errorId)
  })
})
