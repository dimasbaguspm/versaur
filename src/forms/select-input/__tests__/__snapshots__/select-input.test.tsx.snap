// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectInput > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-foreground mb-2"
      for="«r0»"
    >
      Country
    </label>
    <div
      class="relative w-full"
    >
      <select
        aria-invalid="false"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:pointer-events-none disabled:bg-gray-50 appearance-none bg-no-repeat bg-right pr-10 border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 h-9 pl-3"
        id="«r0»"
      >
        <option
          disabled=""
          hidden=""
          value=""
        >
          Select a country
        </option>
        <option
          value="us"
        >
          United States
        </option>
        <option
          value="ca"
        >
          Canada
        </option>
        <option
          value="uk"
        >
          United Kingdom
        </option>
        <option
          value="de"
        >
          Germany
        </option>
        <option
          value="fr"
        >
          France
        </option>
      </select>
      <div
        class="absolute inset-y-0 right-0 flex items-center pr-3 pointer-events-none transition-transform duration-200"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-down inline-flex items-center justify-center w-4 h-4 text-gray-500"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
